File: packages/laipe-vue/src/console.ts
// Debug console — frontend library.
//
// Mirrors the Rust `console.rs` in apps that have a Tauri backend
// (laipe-app). When running browser-only (no Tauri), all features
// still work — the `console:entry` listener and `get_console_entries`
// invoke just no-op / fail silently. Frontend `console.log/warn/error`
// hooks continue to populate the panel.
//
// ## Diagnostic context (v0.2+)
//
// `ConsoleEntry` now carries optional diagnostic fields (`conversation_id`,
// `turn`, `kind`, `request_digest`, `response_digest`, `cause`). The
// chat command fills these when surfacing an LLM error, and the
// `saveReport()` API uses them to locate the on-disk recording and
// synthesize a self-contained `.md` report for an LLM assistant.
//
// Usage:
//   import { initConsole, useConsoleEntries, clearConsole, installConsoleHook, saveReport } from "laipe-vue";
//
//   // In main.ts (once):
//   installConsoleHook();
//   await initConsole();
//
//   // In a component:
//   const entries = useConsoleEntries();
//   entries.value.forEach(...)
//   await saveReport(entries.value[0].id);   // writes <log_dir>/reports/<ts>-<id>.md

import { ref, type Ref } from "vue";
import type { UnlistenFn } from "@tauri-apps/api/event";

export type ConsoleLevel = "info" | "warn" | "error";
export type ConsoleSource = "backend" | "frontend";

/** 8-way error classification. Mirrors `laipe_core::ChatErrorKind`. */
export type ChatErrorKind =
  | "network"
  | "auth"
  | "model_not_found"
  | "bad_request"
  | "rate_limit"
  | "server_error"
  | "stream_protocol"
  | "unknown";

/** Single console entry. Mirrors Rust `ConsoleEntry` (snake_case via serde). */
export interface ConsoleEntry {
  id: string;
  level: ConsoleLevel;
  source: ConsoleSource;
  module: string;
  message: string;
  /** Milliseconds since epoch (snake_case to match Rust). */
  timestamp_ms: number;
  // === Diagnostic context (all optional) ===
  /** Conversation id the entry belongs to. `undefined` for app-wide logs. */
  conversation_id?: string;
  /** Agent-loop turn (0-based). `undefined` outside an agent turn. */
  turn?: number;
  /** Typed error class. Drives the saved report's "Likely causes" section. */
  kind?: ChatErrorKind;
  /** One-line summary of the outgoing request. */
  request_digest?: string;
  /** One-line summary of the incoming response (or error). */
  response_digest?: string;
  /** The original lower-level error string (reqwest / serde / etc.). */
  cause?: string;
}

/** Result of `saveReport`. `path` is relative to the app log dir. */
export interface SavedReport {
  path: string;
}

const MAX_ENTRIES = 1000;

// === Module-level singleton state ===
//
// Console is a single stream — all views share the same data. Module-level
// ref makes the UI auto-reactive without needing a Pinia store.
const entries: Ref<ConsoleEntry[]> = ref([]);
let initialSnapshotLoaded = false;
let unsubscribeEvent: UnlistenFn | null = null;
let consoleHookInstalled = false;

function pushLocal(entry: ConsoleEntry): void {
  // Dedup by id (defends against the listen + snapshot race).
  if (entries.value.some((e) => e.id === entry.id)) return;
  entries.value = [entry, ...entries.value];
  if (entries.value.length > MAX_ENTRIES) {
    entries.value = entries.value.slice(0, MAX_ENTRIES);
  }
}

function formatArgs(args: unknown[]): string {
  return args
    .map((a) => {
      if (typeof a === "string") return a;
      if (a instanceof Error) return a.stack || a.message;
      try {
        return JSON.stringify(a, null, 2);
      } catch {
        return String(a);
      }
    })
    .join(" ");
}

function pushFrontend(level: ConsoleLevel, args: unknown[]): void {
  pushLocal({
    id: `fe-${Date.now()}-${Math.random().toString(36).slice(2, 8)}`,
    level,
    source: "frontend",
    module: "app",
    message: formatArgs(args),
    timestamp_ms: Date.now(),
  });
}

async function loadInitialSnapshot(): Promise<void> {
  if (initialSnapshotLoaded) return;
  try {
    const { invoke } = await import("@tauri-apps/api/core");
    const fromBackend = await invoke<ConsoleEntry[]>("get_console_entries");
    for (const e of fromBackend) pushLocal(e);
    initialSnapshotLoaded = true;
  } catch {
    // No Tauri runtime, or command not registered. Silent.
  }
}

async function subscribeBackendEvents(): Promise<void> {
  if (unsubscribeEvent) return;
  try {
    const { listen } = await import("@tauri-apps/api/event");
    unsubscribeEvent = await listen<ConsoleEntry>("console:entry", (e) => {
      pushLocal(e.payload);
    });
  } catch {
    // No Tauri runtime. Silent.
  }
}

/**
 * Hook `console.log` / `console.warn` / `console.error` so every
 * frontend log line also lands in the debug console. Install once
 * (idempotent) at app startup.
 */
export function installConsoleHook(): void {
  if (consoleHookInstalled) return;
  consoleHookInstalled = true;

  const origLog = console.log.bind(console);
  const origWarn = console.warn.bind(console);
  const origError = console.error.bind(console);

  console.log = (...args: unknown[]) => {
    origLog(...args);
    pushFrontend("info", args);
  };
  console.warn = (...args: unknown[]) => {
    origWarn(...args);
    pushFrontend("warn", args);
  };
  console.error = (...args: unknown[]) => {
    origError(...args);
    pushFrontend("error", args);
  };
}

/** Pull snapshot + subscribe to backend events. */
export async function initConsole(): Promise<void> {
  await Promise.all([loadInitialSnapshot(), subscribeBackendEvents()]);
}

/** Clear all console entries (frontend + backend). */
export async function clearConsole(): Promise<void> {
  entries.value = [];
  try {
    const { invoke } = await import("@tauri-apps/api/core");
    await invoke("clear_console");
  } catch {
    // No Tauri runtime. Frontend-only clear already happened.
  }
}

/** Reactive ref to the entries list. Read this in your console UI. */
export function useConsoleEntries(): Ref<ConsoleEntry[]> {
  return entries;
}

/** Force reload of the backend snapshot (UI "Refresh" button). */
export async function refreshConsole(): Promise<void> {
  initialSnapshotLoaded = false;
  await loadInitialSnapshot();
}

/**
 * Synthesize a self-contained `.md` error report for one console
 * entry. Backend locates the on-disk recording by
 * `(conversation_id, turn)`, assembles the YAML-frontmatter `.md`,
 * and appends a one-line entry to `INDEX.jsonl`.
 *
 * Returns the path to the report (relative to the app log dir).
 * Throws if the entry has no matching recording (e.g. the entry
 * was emitted before diagnostics initialized, or the user manually
 * cleared the recordings dir).
 *
 * Usage from a Vue component:
 * ```ts
 * import { saveReport } from "laipe-vue";
 * const path = await saveReport(entry.id);
 * showToast(`Report saved: ${path}`);
 * ```
 */
export async function saveReport(consoleId: string): Promise<SavedReport> {
  const { invoke } = await import("@tauri-apps/api/core");
  const path = await invoke<string>("dump_error_report", {
    consoleId,
  });
  return { path };
}

/**
 * Diagnostic mode config. Mirrors the Rust `DiagnosticConfig` in
 * `laipe-app/src-tauri/src/diagnostics.rs`. Frontend apps expose
 * these as toggles in Settings; the backend uses them to decide
 * whether to record every round / auto-snapshot errors.
 */
export interface DiagnosticConfig {
  /** Auto-snapshot every error to a `.md` report (off by default). */
  auto_snapshot: boolean;
  /** Max bytes per saved report (response is truncated beyond this). */
  max_report_bytes: number;
  /** Record every chat round, not just failures (off by default). */
  record_successful_rounds: boolean;
}

/** Get the current diagnostic config. */
export async function getDiagnosticConfig(): Promise<DiagnosticConfig> {
  const { invoke } = await import("@tauri-apps/api/core");
  return invoke<DiagnosticConfig>("get_diagnostic_mode");
}

/** Set the diagnostic config. Pass the full object — partial updates
 * are not supported. */
export async function setDiagnosticConfig(cfg: DiagnosticConfig): Promise<void> {
  const { invoke } = await import("@tauri-apps/api/core");
  await invoke("set_diagnostic_mode", { cfg });
}

File: laipe-app/src-tauri/src/console.rs
//! Debug console — in-memory log buffer + Tauri event + commands.
//!
//! Pattern borrowed from PlotCraft's `console.rs` (AGENTS.md 硬规则 #1：结构对齐，代码自写).
//! Simplifications for laipe-starter:
//! - 3 levels (info / warn / error), 2 sources (backend / frontend)
//! - No automatic tracing hook (key error points call `console_log()` manually)
//! - In-memory only, max 1000 entries, FIFO truncation
//!
//! ## Diagnostic context (v0.2+)
//!
//! `ConsoleEntry` now carries LLM-debuggable context fields. All are
//! optional so existing call sites that just want a plain log line
//! don't have to fill them in. The chat command fills them when it
//! surfaces an error so the saved-error-report flow has everything
//! it needs in one round-trip.
//!
//! Data flow:
//! - Rust: `console::console_log(app, "error", "llm", msg)` → push + emit `console:entry`
//! - Frontend: `listen<ConsoleEntry>("console:entry", ...)` accumulates
//! - Frontend init: `invoke("get_console_entries")` for snapshot
//! - UI: `useConsoleEntries()` reactive list, filter / search / clear

use laipe_core::ChatErrorKind;
use serde::Serialize;
use std::sync::Mutex;
use tauri::{AppHandle, Emitter, Manager};

/// One console log entry — frontend mirrors this shape exactly.
#[derive(Debug, Clone, Serialize)]
pub struct ConsoleEntry {
    /// Unique id (timestamp + counter, sufficient for in-memory dedup).
    pub id: String,
    /// Log level: "info" | "warn" | "error".
    pub level: String,
    /// Source: "backend" | "frontend".
    pub source: String,
    /// Module / category name (e.g. "llm", "settings", "app").
    pub module: String,
    /// The log message.
    pub message: String,
    /// Timestamp in milliseconds since epoch.
    pub timestamp_ms: i64,
    // === Diagnostic context (all optional; see `push_with_diag`) ===
    /// Conversation id the entry belongs to. `None` for app-wide logs.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub conversation_id: Option<String>,
    /// Agent-loop turn (0-based). `None` for logs outside an agent turn.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub turn: Option<u32>,
    /// Typed error class. Drives the saved report's "Likely causes"
    /// section via `ChatErrorKind::to_debug_hint()`. `None` for non-error
    /// entries or errors that don't fit the 8-way taxonomy.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub kind: Option<String>,
    /// One-line summary of the outgoing request. Mirrors what the
    /// diagnostic recorder's `meta.json` has. Saved report includes
    /// the full body; this is the in-memory digest.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub request_digest: Option<String>,
    /// One-line summary of the incoming response (or error). The saved
    /// report's `response.bin` is the full raw bytes; this is the digest.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub response_digest: Option<String>,
    /// The original lower-level error string (reqwest / serde / etc.)
    /// the streaming layer classified. Useful for stack-style reading.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub cause: Option<String>,
}

/// Tauri-managed in-memory console buffer.
pub struct ConsoleState {
    entries: Mutex<Vec<ConsoleEntry>>,
    max_entries: usize,
    counter: Mutex<u64>,
}

impl Default for ConsoleState {
    fn default() -> Self {
        Self::new()
    }
}

impl ConsoleState {
    pub fn new() -> Self {
        Self {
            entries: Mutex::new(Vec::new()),
            max_entries: 1000,
            counter: Mutex::new(0),
        }
    }

    /// Push a backend entry. Newest at index 0; FIFO truncation; emit event.
    pub fn push(&self, app: &AppHandle, level: &str, module: &str, message: impl Into<String>) {
        self.push_with_diag(app, level, module, message, ConsoleDiag::default());
    }

    /// Push a backend entry with diagnostic context. All `ConsoleDiag`
    /// fields are optional; the empty `ConsoleDiag` is the same as `push`.
    pub fn push_with_diag(
        &self,
        app: &AppHandle,
        level: &str,
        module: &str,
        message: impl Into<String>,
        diag: ConsoleDiag,
    ) {
        let id = {
            let mut c = self.counter.lock().expect("console counter mutex poisoned");
            *c += 1;
            format!("console-{}-{}", chrono::Utc::now().timestamp_millis(), *c)
        };
        let entry = ConsoleEntry {
            id,
            level: level.to_string(),
            source: "backend".to_string(),
            module: module.to_string(),
            message: message.into(),
            timestamp_ms: chrono::Utc::now().timestamp_millis(),
            conversation_id: diag.conversation_id,
            turn: diag.turn,
            kind: diag.kind.map(|k| k.as_str().to_string()),
            request_digest: diag.request_digest,
            response_digest: diag.response_digest,
            cause: diag.cause,
        };
        {
            let mut entries = self.entries.lock().expect("console entries mutex poisoned");
            entries.insert(0, entry.clone());
            if entries.len() > self.max_entries {
                entries.truncate(self.max_entries);
            }
        }
        // Emit — failure is non-fatal (Tauri runtime may not be up yet).
        let _ = app.emit("console:entry", &entry);
    }

    pub fn snapshot(&self) -> Vec<ConsoleEntry> {
        self.entries
            .lock()
            .expect("console entries mutex poisoned")
            .clone()
    }

    pub fn clear(&self) {
        self.entries
            .lock()
            .expect("console entries mutex poisoned")
            .clear();
    }
}

/// Optional diagnostic context for a `ConsoleEntry`. Builder-style
/// construction keeps call sites readable:
///
/// ```ignore
/// ConsoleDiag::new()
///     .with_conversation_id(conv_id)
///     .with_turn(turn)
///     .with_kind(ChatErrorKind::Auth)
///     .with_request_digest("model=gpt-4o format=openai_chat")
///     .with_response_digest("status=401 content-type=application/json")
///     .with_cause("HTTP status client error (401) for url (...)")
/// ```
#[derive(Debug, Clone, Default)]
pub struct ConsoleDiag {
    pub conversation_id: Option<String>,
    pub turn: Option<u32>,
    pub kind: Option<ChatErrorKind>,
    pub request_digest: Option<String>,
    pub response_digest: Option<String>,
    pub cause: Option<String>,
}

impl ConsoleDiag {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn with_conversation_id(mut self, id: impl Into<String>) -> Self {
        self.conversation_id = Some(id.into());
        self
    }

    pub fn with_turn(mut self, turn: u32) -> Self {
        self.turn = Some(turn);
        self
    }

    pub fn with_kind(mut self, kind: ChatErrorKind) -> Self {
        self.kind = Some(kind);
        self
    }

    pub fn with_request_digest(mut self, digest: impl Into<String>) -> Self {
        self.request_digest = Some(digest.into());
        self
    }

    pub fn with_response_digest(mut self, digest: impl Into<String>) -> Self {
        self.response_digest = Some(digest.into());
        self
    }

    pub fn with_cause(mut self, cause: impl Into<String>) -> Self {
        self.cause = Some(cause.into());
        self
    }
}

/// Helper for any Rust code that wants to log to the console.
/// Resolves `ConsoleState` from the `AppHandle` and pushes an entry.
///
/// Usage:
/// ```ignore
/// console::log(&app, "error", "llm", format!("start_chat failed: {e}"));
/// ```
pub fn log(app: &AppHandle, level: &str, module: &str, message: impl Into<String>) {
    let state = app.state::<ConsoleState>();
    state.push(app, level, module, message);
}

/// Helper for diagnostic-context pushes. Same as `log` but with `ConsoleDiag`.
///
/// Usage:
/// ```ignore
/// console::log_with_diag(
///     &app, "error", "llm", format!("..."),
///     ConsoleDiag::new()
///         .with_kind(ChatErrorKind::Auth)
///         .with_request_digest(format!("model={}", cfg.model)),
/// );
/// ```
pub fn log_with_diag(
    app: &AppHandle,
    level: &str,
    module: &str,
    message: impl Into<String>,
    diag: ConsoleDiag,
) {
    let state = app.state::<ConsoleState>();
    state.push_with_diag(app, level, module, message, diag);
}

/// Tauri command: pull a full console snapshot (frontend first open).
#[tauri::command]
pub fn get_console_entries(state: tauri::State<ConsoleState>) -> Vec<ConsoleEntry> {
    state.snapshot()
}

/// Tauri command: clear all console entries.
#[tauri::command]
pub fn clear_console(state: tauri::State<ConsoleState>) {
    state.clear();
}

/// Tauri command: look up a single entry by id. Used by the
/// `dump_error_report` flow so the report can correlate the saved
/// recording with the console entry that triggered the save.
#[tauri::command]
pub fn get_console_entry_by_id(
    state: tauri::State<ConsoleState>,
    id: String,
) -> Option<ConsoleEntry> {
    state.snapshot().into_iter().find(|e| e.id == id)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn console_diag_builder() {
        let d = ConsoleDiag::new()
            .with_conversation_id("conv-1")
            .with_turn(2)
            .with_kind(ChatErrorKind::Auth)
            .with_request_digest("model=gpt-4o")
            .with_response_digest("status=401")
            .with_cause("HTTP 401");
        assert_eq!(d.conversation_id.as_deref(), Some("conv-1"));
        assert_eq!(d.turn, Some(2));
        assert_eq!(d.kind, Some(ChatErrorKind::Auth));
        assert_eq!(d.request_digest.as_deref(), Some("model=gpt-4o"));
        assert_eq!(d.response_digest.as_deref(), Some("status=401"));
        assert_eq!(d.cause.as_deref(), Some("HTTP 401"));
    }

    #[test]
    fn empty_diag_is_default() {
        let d = ConsoleDiag::default();
        assert!(d.conversation_id.is_none());
        assert!(d.turn.is_none());
        assert!(d.kind.is_none());
        assert!(d.request_digest.is_none());
        assert!(d.response_digest.is_none());
        assert!(d.cause.is_none());
    }
}
